File: poseidon/poseidon.py
from .constants import *

p = 21888242871839275222246405745257275088548364400416034343698204186575808495617
N_ROUNDS_P = [56, 57, 56, 60, 60, 63, 64, 63, 60, 66, 60, 65, 70, 60, 64, 68]

def sigma(inpt):
    return pow(inpt, 5, p)

def ark(t, C, r, inpt):
    return [(inpt[i] + C[i+r]) % p for i in range(t)]

def mix(t, M, inpt):
    out = [0] * t
    for i in range(t):
        lc = 0
        for j in range(t):
            lc += M[j][i] * inpt[j] % p
        out[i] = lc % p
    return out

def mixLast(t, M, s, inpt):
    lc = 0
    for j in range(t):
        lc += M[j][s] * inpt[j] % p
    return lc % p

def mixS(t, S, r, inpt):
    out = [0] * t
    lc = 0
    for i in range(t):
        lc += S[(t * 2 - 1) * r + i] * inpt[i] % p
    out[0] = lc % p
    for i in range(1, t):
        out[i] = (inpt[i] + inpt[0] * S[(t*2-1)*r + t + i - 1]) % p
    return out

def poseidonEx(nins, nouts, ins, initialState):

    # set constants
    t = nins + 1
    nRoundsF = 8
    nRoundsP = N_ROUNDS_P[t - 2]
    C = POSEIDON_C(t)
    S = POSEIDON_S(t)
    M = POSEIDON_M(t)
    P = POSEIDON_P(t)

    # set holders
    sigmaF = [0] * t
    mixLast_ = [0] * nouts
    mix_ = [0] * t
    mixS_ = [0] * t
    
    # commonly used constant
    hnRoundsF = nRoundsF // 2

    ark_= ark(t, C, 0, [initialState] + ins[:(t-1)])
    for r in range(hnRoundsF-1):
        for j in range(t):
            if r == 0:
                sigmaF[j] = sigma(ark_[j])
            else:
                sigmaF[j] = sigma(mix_[j])
        
        ark_ = ark(t, C, (r+1)*t, sigmaF)
        mix_ = mix(t, M, ark_)
    
    for j in range(t):
        sigmaF[j] = sigma(mix_[j])

    ark_ = ark(t, C, hnRoundsF*t, sigmaF)
    mix_ = mix(t, P, ark_)

    for r in range(nRoundsP):
        if r == 0:
            sigmaP = sigma(mix_[0])
        else:
            sigmaP = sigma(mixS_[0])
        
        inpt = [0] * t
        for j in range(t):
            if j == 0:
                inpt[j] = sigmaP + C[(hnRoundsF+1)*t + r]
            else:
                if r == 0:
                    inpt[j] = mix_[j]
                else:
                    inpt[j] = mixS_[j]
        mixS_ = mixS(t, S, r, inpt)
    
    for r in range(hnRoundsF-1):
        for j in range(t):
            if r == 0:
                sigmaF[j] = sigma(mixS_[j])
            else:
                sigmaF[j] = sigma(mix_[j])

        ark_= ark(t, C, (hnRoundsF+1)*t + nRoundsP + r*t, sigmaF)
        mix_ = mix(t, M, ark_)
    
    for j in range(t):
        sigmaF[j] = sigma(mix_[j])

    for i in range(nouts):
        mixLast_[i] = mixLast(t, M, i, sigmaF)
    
    return mixLast_

def poseidon(*ins):
    return poseidonEx(len(ins), 1, list(ins), 0)[0]

File: tsutil/poseidon/poseidon.ts
import BN from "bn.js"
import { 
    p, N_ROUNDS_P, POSEIDON_C, POSEIDON_S, POSEIDON_M, POSEIDON_P
  } from "./constants"

const pBN = new BN(p)

function zeros(dim: number): BN[] {
    const ar: BN[] = [];
    for (let i = 0; i < dim; i++) 
        ar.push(new BN(0))
    return ar
}

export function sigma(inpt: BN): BN {
    return inpt
        .pow(new BN(5))
        .mod(pBN)
}

export function ark(
    t: number, C: BN[], r: number, inpt: BN[]
): BN[] {
    const out: BN[] = zeros(t)
    for (let i = 0; i < t; i++) 
        out[i] = inpt[i]
            .add(C[i+r])
            .mod(pBN)
    return out
}

export function mix(
    t: number, M: BN[][], inpt: BN[]
): BN[] {
    const out: BN[] = zeros(t)
    for (let i = 0; i < t; i++)
        for (let j = 0; j < t; j++) 
            out[i] = out[i]
                .add(M[j][i].mul(inpt[j]))
                .mod(pBN)
    return out
}

export function mixLast(
    t: number, M: BN[][], s: number, inpt: BN[]
): BN {
    let lc: BN = new BN(0)
    for (let j = 0; j < t; j++) 
        lc = lc
            .add(M[j][s].mul(inpt[j]))
            .mod(pBN)
    return lc
}

export function mixS(
    t: number, S: BN[], r: number, inpt: BN[]
): BN[] {
    const shift = (t * 2 - 1) * r
    const out: BN[] = zeros(t)
    let lc = new BN(0)
    for (let i = 0; i < t; i++) 
        lc = lc
            .add(S[shift + i].mul(inpt[i]))
            .mod(pBN)
    out[0] = lc
    for (let i = 1; i < t; i++) 
        out[i] = inpt[0]
            .mul(S[shift + t + i - 1])
            .add(inpt[i])
            .mod(pBN)
    return out
}

export function poseidonEx(
    nins: number, nouts: number, ins: BN[], initialState: BN
) {
    // set constants
    const t = nins + 1
    const nRoundsF = 8
    const nRoundsP = N_ROUNDS_P[t-2]
    const C = POSEIDON_C(t)!.map(x => new BN(x))
    const S = POSEIDON_S(t)!.map(x => new BN(x))
    const M = POSEIDON_M(t)!.map(l => l.map(x => new BN(x)))
    const P = POSEIDON_P(t)!.map(l => l.map(x => new BN(x)))

    // set holders
    let sigmaF = zeros(t)
    let mixLast_ = zeros(nouts)
    let mix_ = zeros(t)
    let mixS_ = zeros(t)
    let ark_ = zeros(t)

    // commonly used constant
    const hnRoundsF = Math.floor(nRoundsF / 2)
    const arkshift = (hnRoundsF+1)*t

    ark_ = ark(t, C, 0, [initialState].concat(ins.slice(0,t-1)))
    for (let r = 0; r < hnRoundsF-1; r++) {
        for (let j = 0; j < t; j++) {
            if (r == 0)
                sigmaF[j] = sigma(ark_[j])
            else 
                sigmaF[j] = sigma(mix_[j])
        }
        ark_ = ark(t, C, (r+1)*t, sigmaF)
        mix_ = mix(t, M, ark_)
    }

    for (let j = 0; j < t; j++) 
        sigmaF[j] = sigma(mix_[j])
    
    ark_ = ark(t, C, hnRoundsF*t, sigmaF)
    mix_ = mix(t, P, ark_)

    let sigmaP: BN
    for (let r = 0; r < nRoundsP; r++) {
        if (r == 0) 
            sigmaP = sigma(mix_[0])
        else 
            sigmaP = sigma(mixS_[0])
        
        let inpt = zeros(t)
        for (let j = 0; j < t; j++) {
            if (j == 0)
                inpt[j] = sigmaP
                    .add(C[arkshift+r])
                    .mod(pBN)
            else {
                if (r == 0)
                    inpt[j] = mix_[j]
                else
                    inpt[j] = mixS_[j]
            }
        }
        mixS_ = mixS(t, S, r, inpt)
    }

    for (let r = 0; r < hnRoundsF-1; r++) {
        for (let j = 0; j < t; j++) {
            if (r == 0) 
                sigmaF[j] = sigma(mixS_[j])
            else 
                sigmaF[j] = sigma(mix_[j])
        }
        ark_ = ark(t, C, arkshift + nRoundsP + r*t, sigmaF)
        mix_ = mix(t, M, ark_)
    }

    for (let j = 0; j < t; j++) 
        sigmaF[j] = sigma(mix_[j])
    
    for (let i = 0; i < nouts; i++) 
        mixLast_[i] = mixLast(t, M, i, sigmaF)
    
    return mixLast_
}

export function poseidon(ins: BN[]): BN {
    return poseidonEx(ins.length, 1, ins, new BN(0))[0]
}

export function postreidon(ins: string[]): string {
    return poseidon(ins.map(x => new BN(x))).toString()
}